qiskit_nature/circuit/library/ansaetze/excitation_builder.py

```diff
@@ -0,0 +1,115 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+TODO.
+"""
+
+from typing import List, Tuple
+
+import itertools
+import logging
+
+from qiskit_nature.operators.second_quantization import FermionicOp, SecondQuantizedOp
+
+logger = logging.getLogger(__name__)
+
+
+class ExcitationBuilder:
+    """A factory class to construct excitation operators."""
+
+    @staticmethod
+    def build_excitation_ops(num_excitations: int,
+                             num_spin_orbitals: int,
+                             num_particles: Tuple[int, int],
+                             alpha_spin: bool = True,
+                             beta_spin: bool = True,
+                             pure_spin: bool = False,
+                             ) -> List[SecondQuantizedOp]:
+        """Builds all possible excitation operators with the given number of excitations for the
+        specified number of particles distributed in the number of orbitals.
+
+        This method assumes block-ordered spin-orbitals.
+
+        Args:
+            num_excitations: number of excitations per operator (1 means single excitations, etc.).
+            num_spin_orbitals: number of spin-orbitals.
+            num_particles: number of alpha and beta particles.
+            alpha_spin: boolean flag whether to include alpha-spin excitations.
+            beta_spin: boolean flag whether to include beta-spin excitations.
+            pure_spin: boolean flag whether to include only pure-spin excitations.
```

Comment:
Can you say what pure spin here means. Does it mean preserve spins i.e. alpha->alpha and not flipping them?

Reply:
Pure spin means that we don't have mixed excitations. E.g. if you consider double excitations, you will only get `alpha,alpha` or `beta,beta` excitations, but no excitations where one is an `alpha` and another is a `beta` particle.

As far as I know in the past (and also the way the new code works now) we always preserve the spin during excitations and never allow flipping.

Reply:
Do people/chemists know what pure-spin is. That it seems to be, from what you sau, to bes what was called same spin before, where the text was more understandable to me since it explained things and did not rely on understanding some  term - I did not hence I thought impure (ie not pure) might be flipping the spin. So when I said 'Can you say here what pure spin means' I meant could you improve the docstring to clarify things - maybe like you explained here in the comments

Reply:
Ah I see :sweat_smile: I will update the code later :+1: I can also call it `same_spin` again, in order to be consistent with what we had before. I didn't check what it used to be and simply went with something that popped into my head.

Reply:
I see that the old `same_spin_doubles` argument actually did exactly the opposite of `pure_spin`: it was used to disable the alpha,alpha and beta,beta excitations while always including alpha,beta ones.
`pure_spin` does the opposite: enable/disable alpha,beta.

I will update the code to use `same_spin_doubles` as this was the original intent (support an equivalent set of configuration options).

Reply:
This was actually a little bit more complicated than I initially assumed. I update the code now so it is worth another check.